### PR TITLE
tiles: Add Tiles

### PR DESCRIPTION
**Pull request by xda2023**
## Break Changes

- Renamed `DockItemState` to `PanelState`.
- Renamed `DockItemInfo` to `PanelInfo`.
- Update `cx` type from `WindowContext` to `AppContext` for Panel trait `panel_name` method.

## TODO

- [ ] Support merge panels into TabPanel.
- [ ] Support split panels into Tile.

### CODE REVIEW

File: crates/ui/src/dock/state.rs [hunk 1]
```diff
@@ -65,6 +68,16 @@ pub struct DockItemState {
     pub info: DockItemInfo,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct TilePanelState {
+    pub(crate) panel_state: DockItemState,
+    pub(crate) x: Pixels,
+    pub(crate) y: Pixels,
+    pub(crate) w: Pixels,
+    pub(crate) h: Pixels,
```
Comment (huacnlee) [inline crates/ui/src/dock/state.rs]: `bounds: Bounds<Pixels>`

File: crates/ui/src/dock/state.rs [hunk 2]
```diff
@@ -65,6 +68,16 @@ pub struct DockItemState {
     pub info: DockItemInfo,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct TilePanelState {
+    pub(crate) panel_state: DockItemState,
```
Comment (huacnlee) [inline crates/ui/src/dock/state.rs]: `panels: Vec<DockItemState>`

File: crates/ui/src/dock/tiles_panel.rs [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+    }
+
+    /// Remove panel from the children.
+    pub fn remove_panel(&mut self, panel: Arc<dyn PanelView>, cx: &mut ViewContext<Self>) {
+        if let Some(ix) = self.index_of_panel(panel.clone()) {
+            self.panels.remove(ix);
+
+            cx.emit(PanelEvent::LayoutChanged);
+        } else {
+            println!("Panel not found.");
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: Remove println

File: crates/ui/src/dock/tiles_panel.rs [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+        }
+    }
+
+    fn update_initial_position(
+        &mut self,
+        position: Point<Pixels>,
+        cx: &mut ViewContext<'_, TilesPanel>,
+    ) {
+        if let Some((index, item)) = self.find_panel_at_position(position) {
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]:
```rs
if let Some((index, item)) = self.find_panel_at_position(position) else {
  return;
}
```

For avoid indent block

File: crates/ui/src/dock/tiles_panel.rs [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+        }
+    }
+
+    fn update_position(
+        &mut self,
+        current_mouse_position: Point<Pixels>,
+        cx: &mut ViewContext<'_, TilesPanel>,
+    ) {
+        if let Some(index) = self.dragging_panel_index {
+            if let Some(item) = self.panels.get_mut(index) {
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: Same here:

```rs
if let Some(index) = self.dragging_panel_index else {
  return;
}

if let Some(item) = self.panels.get_mut(index) else {
  return;
}
```

File: crates/ui/src/dock/tiles_panel.rs [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+                let adjusted_position = current_mouse_position - self.bounds.origin;
+                let delta = adjusted_position - self.dragging_initial_mouse;
+                let mut new_origin = self.dragging_initial_bounds.origin + delta;
+                if new_origin.x < px(0.0) {
+                    new_origin.x = px(0.0);
+                }
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]:
```rs
new_origin.x = new_origin.x.min(px(0.0));
```

File: crates/ui/src/dock/tiles_panel.rs [hunk 7]
```diff
@@ -0,0 +1,728 @@
...
+    fn find_panel_at_position(&self, position: Point<Pixels>) -> Option<(usize, &TilesItem)> {
+        let adjusted_position = position - self.bounds.origin;
+        let mut panels_with_indices: Vec<(usize, &TilesItem)> =
+            self.panels.iter().enumerate().collect();
+        panels_with_indices.sort_by(|a, b| b.1.z_index.cmp(&a.1.z_index));
+
+        for (index, item) in panels_with_indices {
+            let extended_bounds = Bounds::new(
+                item.bounds.origin,
+                Size {
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]:
```rs
item.bounds.size + gpui::size(HALF_HANDLE_SIZE, HALF_HANDLE_SIZE)
```

File: crates/ui/src/dock/tiles_panel.rs [hunk 8]
```diff
@@ -0,0 +1,728 @@
...
+    }
+}
+impl EventEmitter<PanelEvent> for TilesPanel {}
+impl EventEmitter<DismissEvent> for TilesPanel {}
+
+impl Render for TilesPanel {
+    fn render(&mut self, cx: &mut ViewContext<Self>) -> impl IntoElement {
+        let entity_id = cx.entity_id();
+        let view = cx.view().clone();
+        let border = cx.theme().border;
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: Avoid define a color variable, just use `cx.theme().border` in place where you want.

Then will better for read.

File: crates/ui/src/dock/tiles_panel.rs [hunk 9]
```diff
@@ -0,0 +1,728 @@
...
+                        let is_occluded = |bounds: &Bounds<Pixels>| {
+                            self.is_bounds_occluded(current_index, bounds)
+                        };
+
+                        let right_handle_bounds = Bounds::new(
+                            Point {
+                                x: item.bounds.origin.x + item.bounds.size.width
+                                    - px(HALF_HANDLE_SIZE),
+                                y: item.bounds.origin.y,
+                            },
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]:
```rs
item.bounds.origin + point(item.bounds.size.width - px(HALF_HANDLE_SIZE), px(0.))
```

File: crates/ui/src/dock/tiles_panel.rs [hunk 10]
```diff
@@ -0,0 +1,728 @@
...
+                                    .on_mouse_down(
+                                        MouseButton::Left,
+                                        cx.listener(move |this, event: &MouseDownEvent, cx| {
+                                            let initial_mouse_position = event.position;
+                                            let panel_bounds = item.bounds;
+                                            let drag_data = ResizeDragData {
+                                                axis: ResizeAxis::Horizontal,
+                                                initial_mouse_position,
+                                                initial_panel_bounds: panel_bounds,
+                                            };
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: - `initial_mouse_position` -> `last_position`
- `initial_panel_bounds` -> `last_bounds`

To simplify code

File: crates/ui/src/dock/tiles_panel.rs [hunk 11]
```diff
@@ -0,0 +1,728 @@
...
+                                            this.update_resizing_drag(drag_data, cx);
+                                            this.bring_panel_to_front(this.resizing_panel_index);
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: `resizing_panel_index` -> `resizing_index`

In the `tiles`, the primary item is the `panel` not have any other things.

So `active_index`, `resizing_index` that must means the `panel`.

File: crates/ui/src/dock/tiles_panel.rs [hunk 12]
```diff
@@ -0,0 +1,728 @@
...
+                                                    if let Some(ref drag_data) =
+                                                        this.resizing_drag_data
+                                                    {
+                                                        if drag_data.axis != ResizeAxis::Horizontal
+                                                        {
+                                                            return;
+                                                        }
+                                                        let current_mouse_position =
+                                                            e.event.position;
+                                                        let delta = current_mouse_position.x
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: `current_mouse_position` -> `pos` or `position` to less code

File: crates/ui/src/dock/tiles_panel.rs [hunk 13]
```diff
@@ -0,0 +1,728 @@
...
+                                            }
+                                        },
+                                    ))
+                                    .into_any_element()
+                            } else {
+                                div().into_any_element()
+                            })
+                            .child(if !is_occluded(&bottom_handle_bounds) {
+                                div()
+                                    .id("bottom-resize-handle")
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: Split all `resize-handle` into `render_resize_handles` method.

File: crates/ui/src/dock/tiles_panel.rs [hunk 14]
```diff
@@ -0,0 +1,728 @@
...
+                                            }
+                                        },
+                                    ))
+                                    .into_any_element()
+                            } else {
+                                div().into_any_element()
+                            })
+                            .child(if !is_occluded(&drag_bar_bounds) {
+                                h_flex()
+                                    .id("drag-bar")
```
Comment (huacnlee) [inline crates/ui/src/dock/tiles_panel.rs]: Split drag-bar to `render_drag_bar` method.

File: crates/ui/src/dock/state.rs (hunk 1 above)
Comment (huacnlee) [reply crates/ui/src/dock/state.rs]: GPUI's Bounds, Point how is serializable

https://github.com/zed-industries/zed/pull/21783

Comment (xda2023) [reply crates/ui/src/dock/state.rs]: 46c1562

File: crates/ui/src/dock/tiles_panel.rs (hunk 3 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 28c4534

File: crates/ui/src/dock/tiles_panel.rs (hunk 4 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: f450b1b

File: crates/ui/src/dock/tiles_panel.rs (hunk 5 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 6b0f11c

File: crates/ui/src/dock/tiles_panel.rs (hunk 6 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 11a19e0

File: crates/ui/src/dock/tiles_panel.rs (hunk 7 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 934dfe8

File: crates/ui/src/dock/tiles_panel.rs (hunk 8 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: c8a46bc

File: crates/ui/src/dock/tiles_panel.rs (hunk 9 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 3f434f9

File: crates/ui/src/dock/tiles_panel.rs (hunk 10 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 04a5658

File: crates/ui/src/dock/tiles_panel.rs (hunk 11 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: 3f1cca8

File: crates/ui/src/dock/tiles_panel.rs (hunk 12 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: e4c87ef

File: crates/ui/src/dock/tiles_panel.rs (hunk 13 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: e7558a3

File: crates/ui/src/dock/tiles_panel.rs (hunk 14 above)
Comment (xda2023) [reply crates/ui/src/dock/tiles_panel.rs]: e8e540d

File: crates/ui/src/dock/state.rs (hunk 2 above)
Comment (xda2023) [reply crates/ui/src/dock/state.rs]: 06b5fc2

**Pull request merged**